modules/http4s-kernel/src/smithy4s/http4s/kernel/Md5CheckSumClient.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.http4s.kernel
+
+import smithy4s._
+
+import org.http4s._
+import org.http4s.client.Client
+import cats.effect.Concurrent
+import cats.effect.Resource
+import fs2.{Chunk, Pipe}
+import cats.syntax.all._
+import org.typelevel.ci.CIString
+
+object Md5CheckSumClient {
+  def apply[F[_]: Concurrent](hints: Hints): Client[F] => Client[F] = {
+    client =>
+      hints.get(smithy.api.HttpChecksumRequired) match {
+        case Some(_) =>
+          reqWithChecksum[F](client)
+        case _ => client
+      }
+  }
+
+  private def reqWithChecksum[F[_]: Concurrent](
+      client: Client[F]
+  ): Client[F] = {
+    val md5HeaderPipe: Pipe[F, Byte, String] =
+      fs2.hash.md5[F] andThen fs2.text.base64.encode[F]
+    Client { request =>
+      val bodyF = request.body.through(md5HeaderPipe).compile.to(Chunk)
```

Comment:
**issue:** I think you might need a strict request for this. Otherwise, future attempts to consume the body stream may fail (this might not be a concern for bodies created from plain byte arrays, but it's good practice nevertheless to keep that stream consumable)

Reply:
Should the same apply here https://github.com/disneystreaming/smithy4s/blob/5902d2fd81d1eee6703daaab6e2ad95eb69684ab/modules/aws-http4s/src/smithy4s/aws/internals/AwsSignatureClient.scala#LL94C71-L94C71

Reply:
@yisraelU you are correct, it should indeed. Right now it's worked because we've only wired it to strict requests. I'll fix :)

Reply:
any idea why `Request(..).toStrict` ,requires an instance of Concurrent ?
compiling a stream can accept multiple targets including Sync .

Reply:
I think we could copy-paste the implementation to avoid bringing in `Concurrent`. Up to you.

Reply:
since it's client-side, doing ` val bodyF = request.body.through(md5HeaderPipe).compile.to(Chunk)` and then `withEntityBody` at the end re-using the produced chunk might be more straightforward actually. 

`toStrict` is more useful to protect servers against ill-intended clients.

Reply:
Isn't it possible that this client would be called to proxy up a request on the server side?

Reply:
In the grand scheme of things that's true, but the purpose of this middleware is not to be used with "any" http4s client, it's more about clients that are used under the hood for smithy4s clients, the requests of which is gonna be derived from high-level data types that live fully in memory

Reply:
> since it's client-side, doing ` val bodyF = request.body.through(md5HeaderPipe).compile.to(Chunk)` and then `withEntityBody` at the end re-using the produced chunk might be more straightforward actually.

Did you mean `withEntity` ?
Also, we can't "reuse" the chunk as it's hashed , 
Does it make sense to compile to a F[Chunk[Byte]] and then Stream.emits to pipe it through md5 to create header?

Reply:
That checks out.